aspnet/migration/http-modules/sample/Asp.Net5/src/Asp.Net5/Startup.cs

```diff
@@ -0,0 +1,110 @@
+﻿// ASP.NET 5 Startup class
+
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+using Asp.Net5.Middleware;
+using MyApp.Middleware;
+
+namespace Asp.Net5
+{
+    public class Startup
+    {
+        public Startup(IHostingEnvironment env)
+        {
+            // Set up configuration sources.
+            var builder = new ConfigurationBuilder()
+                .AddJsonFile("appsettings.json")
+                .AddEnvironmentVariables();
+            Configuration = builder.Build();
+        }
+
+        public IConfigurationRoot Configuration { get; set; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Setup options service
+            services.AddOptions();
+
+            // ...
+
+            // Load options from section "MyMiddlewareOptionsSection"
+            services.Configure<MyMiddlewareOptions>(
+                Configuration.GetSection("MyMiddlewareOptionsSection"));
+
+            // ...
+
+            // Add framework services.
+            services.AddMvc();
+        }
+
+        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+        public void Configure(IApplicationBuilder app, IHostingEnvironment env, 
+            ILoggerFactory loggerFactory)
+        {
+            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
+            loggerFactory.AddDebug();
+
+            app.UseDeveloperExceptionPage();
+
+            // ...
+
+            app.UseMyMiddleware();
+
+            // ...
+
+            app.UseMyMiddlewareWithParams();
+
+            var myMiddlewareOptions = 
+                Configuration.Get<MyMiddlewareOptions>("MyMiddlewareOptionsSection");
+
+            var myMiddlewareOptions2 = 
+                Configuration.Get<MyMiddlewareOptions>("MyMiddlewareOptionsSection2");
+
+            app.UseMyMiddlewareWithParams(myMiddlewareOptions);
+
+            // ...
+
+            app.UseMyMiddlewareWithParams(myMiddlewareOptions2);
+
+            app.UseMyTerminatingMiddleware();
+
+            app.UseIISPlatformHandler();
+
+            app.UseSimpleAuthorize();
+
+            // Create branch to the MyHandlerMiddleware. 
+            // All requests ending in .report will follow this branch.
+            app.MapWhen(
```

Comment:
I considered this option as well. I went with the MapWhen alternative because:

1) It decouples the routing of the request from the middleware itself, making it easier to later on route requests for .myreport or .financereport to the same report middleware as well;

2) Having the middleware decide to handle a request based on the url means moving some routing logic from the Configure method into the middleware. So now routing is in two places. Also, to someone simply looking at the Configure method, it wouldn't be obvious that the reporting middleware is potentially a handler (that short circuits requests), rather than a "normal" middleware that has requests passing through.

3) With handlers, the extension associated with a handler is set at the point where it is configured (in web.config), not inside the handler code itself. The Configure method in many ways is the ASP.NET 5 counterpart of web.config - so specifying the extension in the Configure method (using MapWhen) provides a more direct mapping from handlers to middleware.